tools/src/ops_tools/_generate_yaml.py

```diff
@@ -0,0 +1,434 @@
+
+
+class ActionDict(TypedDict, total=False):
+    description: str
+    params: dict[str, Any]
+    """A dictionary of parameters for the action."""
+    required: list[str]
+    """A list of required parameters for the action."""
+    additionalProperties: bool
+    """Whether additional properties are allowed in the action parameters."""
+
+
+JUJU_TYPES: Final[Mapping[type, str]] = {
+    bool: 'boolean',
+    int: 'int',
+    float: 'float',
+    str: 'string',
+    ops.Secret: 'secret',
+}
+
+# We currently only handle the basic types that we expect to see in real charms.
+# lists and tuples (arrays) are handled without using this mapping.
+JSON_TYPES: Final[Mapping[type, str]] = {
+    bool: 'boolean',
+    int: 'integer',
+    float: 'number',
+    str: 'string',
+}
+
+
+def attr_to_default(cls: type[object], name: str) -> Any:
+    """Get the default value for the attribute."""
+    if not dataclasses.is_dataclass(cls):
+        return getattr(cls, name, None)
+    default = None
+    for field in dataclasses.fields(cls):
+        if field.name != name:
+            continue
+        break
+    else:
+        return default
+
+    # This might be a Pydantic dataclass using a Pydantic.Field object.
+    field_default = (  # type: ignore
+        field.default.default  # type: ignore
+        if hasattr(field.default, 'default')
+        else field.default
+    )
+    field_default_factory = (  # type: ignore
+        field.default.default_factory  # type: ignore
+        if hasattr(field.default, 'default_factory')
+        else field.default_factory
+    )
+    # A hack to avoid importing Pydantic here.
+    if (
+        'PydanticUndefinedType' not in str(type(field_default))  # type: ignore
+        and field_default is not dataclasses.MISSING
+    ):
+        return field_default  # type: ignore
+    if field_default_factory is not dataclasses.MISSING:
+        return field_default_factory()  # type: ignore
+    return default
+
+
+def _attr_to_yaml_type(cls: type[object], name: str, yaml_types: dict[type, str]) -> str:
+    try:
+        raw_hint = get_type_hints(cls)[name]
+    except KeyError:
+        pass
+    else:
+        # Collapse Optional[] and Union[] and so on to the simpler form.
+        origin = get_origin(raw_hint)
+        if origin in (list, tuple):
+            return yaml_types[origin]
+        elif origin:
+            hints = {arg for arg in get_args(raw_hint) if arg in yaml_types}
+        else:
+            hints = {raw_hint}
+        # If there are multiple types -- for example, the type annotation is
+        # `int | str` -- then we can't determine the type, and we fall back to
+        # "string", even if `str` is not in the type hint, because our
+        # "we can't determine the type" choice is always "string".
+        if len(hints) > 1:
+            return 'string'
+        elif hints:
+            try:
+                return yaml_types[hints.pop()]
+            except KeyError:
+                pass
```

Comment:
I think this is a simpler way of writing the same logic -- with one difference: we now *always* return `'string'` if there are multiple hints, instead of making an exception for the case where exactly one of them is in `yaml_types`. To bring that logic back, filter the result of `get_args` based on `yaml_types` -- but I think the semantics are clearer this way.

```suggestion
        hints = set(get_args(raw_hint)) if origin else {raw_hint}
        # If there are multiple types -- for example, the type annotation is
        # `int | str` -- then we can't determine the type.
        # Likewise if there are no hints somehow (generic used without args?).
        # If we can't figure it out, then use "string", which should be the most
        # compatible, and most likely to be used for arbitrary types. Charms can
        # override `to_juju_schema` to adjust this if required.
        return yaml_types.get(hints.pop(), 'string') if len(hints) == 1 else 'string'
```